# Functions

In [10]:
lists=[12,15,1,2,3,4,5,6,7,8,9,10]

def analyze_numbers(numbers):
    """function analyze_numbers(numbers) that takes a list of integers and returns a tuple"""
    largest_number=0
    sum_all_number=0
    smallest_number=numbers[0]
    divider=len(numbers)
    for i in numbers:
        if i>largest_number:
            largest_number=i
        if i<smallest_number:
            smallest_number=i
        sum_all_number+=i
    average=sum_all_number//divider
    tup=(largest_number,smallest_number,sum_all_number,average)
    return tup
        

In [11]:
analyze_numbers(lists)


(15, 1, 82, 6)

Write a function second_largest(numbers) that returns the second-largest unique number from a list.

Example:

Input:  [10, 5, 20, 20, 8, 15]
Output: 15

Handle the case where a second-largest unique value doesn't exist.

In [13]:
def second_largest(numbers):
    """function second_largest(numbers) that returns the second-largest unique number from a list"""
    largest_number=0
    second_largest_number=0
    for i in numbers:
        if i>largest_number:
            second_largest_number=largest_number
            largest_number=i
    return second_largest_number
            

In [14]:
lists=[12,15,1,2,3,4,5,6,7,8,9,10]
second_largest(lists)

12

Write a function frequency_report(text) that counts the frequency of every word in a sentence and returns a dictionary.

Requirements:

Case-insensitive
Ignore . , ! ?
Treat "Python" and "python" as the same word

In [26]:
def frequency_report(text):
    text=text.lower()
    punctuation=".,!"
    for char in punctuation:
        text=text.replace(char,"")
    frequency_dict={}
    words=text.split()
    for word in words:
        if word in frequency_dict:
            frequency_dict[word]+=1
        else:
            frequency_dict[word]=1
        
    return frequency_dict



In [27]:
sentence="hi my name is Hars"
frequency_report(sentence)

{'hi': 1, 'my': 1, 'name': 1, 'is': 1, 'hars': 1}

# Lambda

Given:

students = [
    ("Rahul", 78),
    ("Aman", 92),
    ("Jay", 85),
    ("Riya", 92),
    ("Karan", 67)
]

Use a lambda function to sort the students by:

marks in descending order
if marks are equal, name alphabetically

In [31]:
students = [
    ("Rahul", 78),
    ("Aman", 92),
    ("Jay", 85),
    ("Riya", 92),
    ("Karan", 67)
]

ans=sorted(students,key=lambda x:(-x[1],x[0]))
ans

[('Aman', 92), ('Riya', 92), ('Jay', 85), ('Rahul', 78), ('Karan', 67)]

Use map() with a lambda to transform every word into:

(word, length, uppercase_word)

In [34]:
word= ["python", "machine", "learning", "ai", "data"]
ans=list(map(lambda w:(w,len(w),w.upper()),word))
ans

[('python', 6, 'PYTHON'),
 ('machine', 7, 'MACHINE'),
 ('learning', 8, 'LEARNING'),
 ('ai', 2, 'AI'),
 ('data', 4, 'DATA')]

numbers = [-10, 5, -3, 8, 0, -7, 12, 4]

Using a lambda, transform every number according to:

negative → absolute value
zero → 100
positive → square

In [42]:
numbers = [-10, 5, -3, 8, 0, -7, 12, 4]
ans=list(map(lambda x:abs(x) if x<0 else(100 if x==0 else x**2  ),numbers))
ans

[10, 25, 3, 64, 100, 7, 144, 16]

# map

Given:

numbers = [12, 7, 25, 3, 18, 10, 5]

Using map(), transform each number into:

"12 → Even → 144"
"7 → Odd → 49"

In [48]:
numbers = [12, 7, 25, 3, 18, 10, 5]
list(map(lambda x:(x,"even" if x%2==0 else "odd",x**2),numbers))


[(12, 'even', 144),
 (7, 'odd', 49),
 (25, 'odd', 625),
 (3, 'odd', 9),
 (18, 'even', 324),
 (10, 'even', 100),
 (5, 'odd', 25)]

Given:

prices = [100, 250, 499, 1200, 75]

Using map(), apply:

18% GST
then round the final price to 2 decimal places

Return a new list containing the final prices.

In [50]:
prices = [100, 250, 499, 1200, 75]

ans=list(map(lambda x:round((x*18/100),2),prices))
ans

[18.0, 45.0, 89.82, 216.0, 13.5]

Given:

students = [
    {"name": "Jay", "marks": [80, 75, 90]},
    {"name": "Aman", "marks": [65, 70, 72]},
    {"name": "Riya", "marks": [95, 88, 91]}
]

Use map() to create a new list containing:

(name, average_marks, status)

In [61]:
students = [
    {"name": "Jay", "marks": [80, 75, 90]},
    {"name": "Aman", "marks": [65, 70, 72]},
    {"name": "Riya", "marks": [95, 88, 91]}
]
ans=list(map(lambda x:(x["name"],round(sum(x["marks"])/len(x["marks"]),2),"excellent" if round(sum(x["marks"])/len(x["marks"]),2) >= 75 else ("pass" if round(sum(x["marks"])/len(x["marks"]),2)>=40 else "fail")),students))
ans

[('Jay', 81.67, 'excellent'),
 ('Aman', 69.0, 'pass'),
 ('Riya', 91.33, 'excellent')]

Given:

numbers = [12, 15, 7, 22, 30, 9, 18, 41, 50]

Use filter() to keep numbers that are:

even
greater than 15
divisible by 3

In [68]:
numbers = [12, 15, 7, 22, 30, 9, 18, 41, 50]
ans=list(filter(lambda x:x%2==0 and x>15 and x%3==0,numbers))
print(ans)

[30, 18]


Given:

students = [
    ("Jay", 82),
    ("Aman", 45),
    ("Riya", 91),
    ("Karan", 38),
    ("Neha", 76),
    ("Raj", 55)
]

Use filter() to keep only students who:

scored at least 50
and whose name has at least 4 characters.

In [71]:
students = [
    ("Jay", 82),
    ("Aman", 45),
    ("Riya", 91),
    ("Karan", 38),
    ("Neha", 76),
    ("Raj", 55)
]
ans=list(filter(lambda x: x[1]>=50 and len(x[0])>=4,students))
ans

[('Riya', 91), ('Neha', 76)]

Given:

words = [
    "python",
    "AI",
    "machine",
    "data",
    "ML",
    "learning",
    "code",
    "neural"
]

Use filter() to keep words that satisfy all three:

length ≥ 5
contains the letter "a" (case-insensitive)
does not start with "m"

In [72]:
words = [
    "python",
    "AI",
    "machine",
    "data",
    "ML",
    "learning",
    "code",
    "neural"
]
ans=list(filter(lambda x:len(x)>=5 and "a" in x.lower() and not x.lower().startswith("m"),words))
ans

['learning', 'neural']

# Recursive

Write a recursive function digit_sum(n) that returns the sum of all digits of a positive integer.

Example:

Input: 58329
Output: 27

Restriction: Do not convert the number to a string.

In [73]:
def digit_sum(n):
    if n==0:
        return 0

    return (n%10)+digit_sum(n//10)

input_num=568439
output=digit_sum(input_num)
print(output)

35


In [2]:
def factorial(n):
    if n==1:
        return 1
    return n*factorial(n-1)
factorial(6)

720